pkg/urbit/jets/e/de_json.c

```diff
@@ -0,0 +1,238 @@
+/* j/5/de_json.c
+**
+*/
+#include "all.h"
+#include "pdjson.h"
+
+/* custom code for calling parser imported from:
+** https://github.com/skeeto/pdjson
+*/
+
+typedef struct _u3qedj_coll {
+  // collection (list for array, map for object)
+  u3_noun  col;
+  // store context for recursive arrays/objects:
+  //   - pointer to tail for array
+  //   - key for object
+  union {
+    u3_noun* tel;
+    u3_atom  key;
+  };
+} u3qedj_coll;
+
+static u3qedj_coll*
+_push_stack(const u3a_pile *pil_u)
+{
+  u3qedj_coll *res_u = u3a_push(pil_u);
+  u3a_pile_sane(pil_u);
+
+  res_u->col = u3_nul;
+  res_u->key = u3_none;
+
+  return res_u;
+}
+
+static u3qedj_coll*
+_pop_stack(const u3a_pile *pil_u)
+{
+  return u3a_pop(pil_u);
+}
+
+static void
+_close_stack(const u3a_pile *pil_u)
+{
+  while ( c3n == u3a_pile_done(pil_u) ) {
+    u3qedj_coll *tak_u = u3a_peek(pil_u);
+    u3z(tak_u->col);
+    u3a_drop(pil_u);
+  }
+}
+
+static void
+_close(json_stream *sam_u, const u3a_pile *pil_u)
+{
+  _close_stack(pil_u);
+  json_close(sam_u);
+}
+
+static u3_noun
+_parse(u3_atom txt)
+{
+  //
+  // vars
+  //
+
+  u3qedj_coll *tak_u;
+
+  json_allocator loc_u = {u3a_malloc, u3a_realloc, u3a_free};
+  json_stream    sem_u;
+  json_stream*   sam_u = &sem_u;
+
+  u3a_pile  pel_u;
+  u3a_pile *pil_u = &pel_u;
+
+  u3_noun res = u3_none;
+  u3_noun val;
+
+  const c3_y *byt_y;
+  c3_d        cnt_d;
+  c3_w        len_w = u3r_met(3, txt);
+
+  //
+  // initialization
+  //
+
+  // XX assumes little-endian
+  //
+  if ( c3y == u3a_is_cat(txt) ) {
+    byt_y = (c3_y*)&txt;
+  }
+  else {
+    u3a_atom* vat_u = u3a_to_ptr(txt);
+    byt_y = (c3_y*)vat_u->buf_w;
+  }
+  json_open_buffer(sam_u, byt_y, len_w);
+  json_set_allocator(sam_u, &loc_u);
+  u3a_pile_prep(pil_u, sizeof(u3qedj_coll));
+
+  //
+  // core logic
+  //
+
+  while ( json_peek(sam_u) != JSON_DONE ) {
+    switch ( json_next(sam_u) ) {
+      // unreachable barring programming error
+      default: c3_assert(0);
+
+      case JSON_ARRAY:
+      case JSON_OBJECT: {
+        tak_u = _push_stack(pil_u);
+      } continue;
+
+      case JSON_ARRAY_END: {
+        val = u3nc(c3__a, tak_u->col);
+        tak_u = _pop_stack(pil_u);
+      } break;
+
+      case JSON_OBJECT_END: {
+        val = u3nc(c3__o, tak_u->col);
+        tak_u = _pop_stack(pil_u);
+      } break;
+
+      case JSON_STRING: {
+        size_t      len_i;
+        const c3_c *str_c = json_get_string(sam_u, &len_i);
+        u3_atom     str = (len_i <= 1) ?
+                            u3_nul :
+                            u3i_bytes(len_i - 1, (const c3_y *)str_c);
+
+        if ( (json_get_context(sam_u, &cnt_d) == JSON_OBJECT) && (cnt_d & 1) ) {
+          tak_u->key = str;
+        }
+        else {
+          val = u3nc(c3__s, str);
+        }
```

Comment:
I think the `continue` statement is a good idea because it captures the idea that we're expecting a pair, and because it lets us skip the conditional check for `JSON_OBJECT` later. However, I don't think that it makes the stack relationship clearer - I think you and I just understand that because of looking at this code for so long.

What's the policy on `assert` statements? My understanding is that most C projects use them only during development and as default cases in control flow situations. Said another way: they shouldn't be used outside of control flow in production code to guard against illegal states; if they are, it's a sign that the state transition is poorly defined. I'm asking because if that's the case, then most of the assertions in these jets should be removed once we're done testing, or at least after they've been in production for a while.

I'm my above understanding is incorrect: do you know what the impact of assertions is on performance? I'm having a difficult time finding good numbers and don't want to run a test myself, so I'm wondering if you know off the top of your head.